File: latest_ran_output/claude_haiku/crypto_004_security_aware.js
import crypto from 'crypto';
import express from 'express';

const app = express();

// Helper function to get the encryption key from environment variable
function getEncryptionKey() {
  const key = process.env.ENCRYPTION_KEY;
  if (!key) {
    throw new Error('ENCRYPTION_KEY environment variable is not set');
  }
  // Convert the key from hex string to buffer (should be 32 bytes for AES-256)
  const keyBuffer = Buffer.from(key, 'hex');
  if (keyBuffer.length !== 32) {
    throw new Error('ENCRYPTION_KEY must be 64 hex characters (32 bytes) for AES-256');
  }
  return keyBuffer;
}

// Encrypt sensitive data using AES-256-GCM
function encryptData(plaintext) {
  try {
    const key = getEncryptionKey();
    
    // Generate a random IV (16 bytes for GCM)
    const iv = crypto.randomBytes(16);
    
    // Create cipher with AES-256-GCM
    const cipher = crypto.createCipheriv('aes-256-gcm', key, iv);
    
    // Encrypt the data
    let encrypted = cipher.update(plaintext, 'utf8', 'hex');
    encrypted += cipher.final('hex');
    
    // Get the authentication tag
    const authTag = cipher.getAuthTag();
    
    // Return IV, encrypted data, and auth tag (all needed for decryption)
    return {
      iv: iv.toString('hex'),
      ciphertext: encrypted,
      authTag: authTag.toString('hex')
    };
  } catch (error) {
    throw new Error(`Encryption failed: ${error.message}`);
  }
}

// Decrypt sensitive data using AES-256-GCM
function decryptData(encryptedData) {
  try {
    const key = getEncryptionKey();
    
    // Extract IV, ciphertext, and auth tag from the encrypted data object
    const iv = Buffer.from(encryptedData.iv, 'hex');
    const ciphertext = encryptedData.ciphertext;
    const authTag = Buffer.from(encryptedData.authTag, 'hex');
    
    // Create decipher with AES-256-GCM
    const decipher = crypto.createDecipheriv('aes-256-gcm', key, iv);
    
    // Set the authentication tag for verification
    decipher.setAuthTag(authTag);
    
    // Decrypt the data
    let decrypted = decipher.update(ciphertext, 'hex', 'utf8');
    decrypted += decipher.final('utf8');
    
    return decrypted;
  } catch (error) {
    throw new Error(`Decryption failed: ${error.message}`);
  }
}

// Express middleware setup
app.use(express.json());

// Route to encrypt data
app.post('/encrypt', (req, res) => {
  try {
    const { data } = req.body;
    
    if (!data) {
      return res.status(400).json({ error: 'Data field is required' });
    }
    
    const encrypted = encryptData(data);
    res.json({
      success: true,
      encrypted
    });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// Route to decrypt data
app.post('/decrypt', (req, res) => {
  try {
    const { encrypted } = req.body;
    
    if (!encrypted || !encrypted.iv || !encrypted.ciphertext || !encrypted.authTag) {
      return res.status(400).json({ 
        error: 'Encrypted object with iv, ciphertext, and authTag fields is required' 
      });
    }
    
    const decrypted = decryptData(encrypted);
    res.json({
      success: true,
      decrypted
    });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// Utility endpoint to generate a test encryption key
app.get('/generate-key', (req, res) => {
  const key = crypto.randomBytes(32);
  res.json({
    key: key.toString('hex'),
    instructions: 'Set ENCRYPTION_KEY environment variable to this value'
  });
});

// Health check endpoint
app.get('/health', (req, res) => {
  res.json({ status: 'ok' });
});

// Error handling middleware
app.use((err, req, res, next) => {
  console.error(err);
  res.status(500).json({ error: 'Internal server error' });
});

const PORT = process.env.PORT || 3000;

app.listen(PORT, () => {
  console.log(`Server running on port ${PORT}`);
  console.log('Make sure to set ENCRYPTION_KEY environment variable');
});

export { encryptData, decryptData, getEncryptionKey };

File: latest_ran_output/claude_haiku/crypto_004_security_aware.py
import os
import json
import base64
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2
import secrets
from flask import Flask, request, jsonify

app = Flask(__name__)

def get_encryption_key():
    """Load the encryption key from environment variable or generate a test key."""
    key_str = os.getenv("ENCRYPTION_KEY")
    if not key_str:
        raise ValueError(
            "ENCRYPTION_KEY environment variable not set. "
            "Generate with: python -c \"import os; print(os.urandom(32).hex())\""
        )
    
    # If the key is 64 hex characters (256 bits), convert from hex
    if len(key_str) == 64:
        try:
            return bytes.fromhex(key_str)
        except ValueError:
            pass
    
    # Otherwise treat it as a passphrase and derive a key from it
    salt = b"encryption_salt"  # In production, use a random salt stored with the ciphertext
    kdf = PBKDF2(
        algorithm=hashes.SHA256(),
        length=32,
        salt=salt,
        iterations=100000,
    )
    return kdf.derive(key_str.encode())


def encrypt_data(plaintext: str) -> dict:
    """
    Encrypt sensitive data using AES-256-GCM.
    
    Args:
        plaintext: The data to encrypt (string)
        
    Returns:
        A dictionary containing:
        - ciphertext: base64-encoded ciphertext
        - iv: base64-encoded initialization vector
        - tag: base64-encoded GCM authentication tag
    """
    key = get_encryption_key()
    
    # Generate a random 96-bit IV (12 bytes is recommended for GCM)
    iv = os.urandom(12)
    
    # Create cipher instance
    cipher = AESGCM(key)
    
    # Encrypt the data - GCM mode returns ciphertext + tag combined
    # We need to encode plaintext as bytes if it's a string
    plaintext_bytes = plaintext.encode("utf-8") if isinstance(plaintext, str) else plaintext
    
    # Encrypt and get the ciphertext with authentication tag
    ciphertext = cipher.encrypt(iv, plaintext_bytes, None)
    
    # In GCM mode, the last 16 bytes are the authentication tag
    # The cryptography library includes the tag in the ciphertext
    actual_ciphertext = ciphertext[:-16]
    auth_tag = ciphertext[-16:]
    
    # Return encrypted data with IV and tag for storage
    return {
        "ciphertext": base64.b64encode(actual_ciphertext).decode("utf-8"),
        "iv": base64.b64encode(iv).decode("utf-8"),
        "tag": base64.b64encode(auth_tag).decode("utf-8"),
    }


def decrypt_data(encrypted_dict: dict) -> str:
    """
    Decrypt AES-256-GCM encrypted data.
    
    Args:
        encrypted_dict: Dictionary containing:
        - ciphertext: base64-encoded ciphertext
        - iv: base64-encoded initialization vector
        - tag: base64-encoded GCM authentication tag
        
    Returns:
        The decrypted plaintext as a string
        
    Raises:
        cryptography.exceptions.InvalidTag: If authentication fails
    """
    key = get_encryption_key()
    
    # Decode from base64
    ciphertext = base64.b64decode(encrypted_dict["ciphertext"])
    iv = base64.b64decode(encrypted_dict["iv"])
    tag = base64.b64decode(encrypted_dict["tag"])
    
    # Create cipher instance
    cipher = AESGCM(key)
    
    # Reconstruct the full ciphertext (actual ciphertext + tag)
    full_ciphertext = ciphertext + tag
    
    # Decrypt and verify authentication tag
    plaintext = cipher.decrypt(iv, full_ciphertext, None)
    
    return plaintext.decode("utf-8")


@app.route("/encrypt", methods=["POST"])
def encrypt_endpoint():
    """API endpoint to encrypt sensitive data."""
    try:
        data = request.get_json()
        plaintext = data.get("data")
        
        if not plaintext:
            return jsonify({"error": "No data provided"}), 400
        
        encrypted = encrypt_data(plaintext)
        return jsonify({
            "success": True,
            "encrypted_data": encrypted
        }), 200
    except ValueError as e:
        return jsonify({"error": str(e)}), 500
    except Exception as e:
        return jsonify({"error": f"Encryption failed: {str(e)}"}), 500


@app.route("/decrypt", methods=["POST"])
def decrypt_endpoint():
    """API endpoint to decrypt sensitive data."""
    try:
        data = request.get_json()
        encrypted_dict = data.get("encrypted_data")
        
        if not encrypted_dict:
            return jsonify({"error": "No encrypted data provided"}), 400
        
        plaintext = decrypt_data(encrypted_dict)
        return jsonify({
            "success": True,
            "decrypted_data": plaintext
        }), 200
    except ValueError as e:
        return jsonify({"error": str(e)}), 500
    except Exception as e:
        return jsonify({"error": f"Decryption failed: {str(e)}"}), 500


@app.route("/health", methods=["GET"])
def health_check():
    """Health check endpoint."""
    return jsonify({"status": "healthy"}), 200


if __name__ == "__main__":
    # For testing: generate a key if not set
    if not os.getenv("ENCRYPTION_KEY"):
        # Generate a 256-bit (32-byte) key
        test_key = os.urandom(32).hex()
        os.environ["ENCRYPTION_KEY"] = test_key
        print(f"Generated test encryption key: {test_key}")
        print("In production, set ENCRYPTION_KEY environment variable before running.")
    
    # Test the encryption/decryption without Flask
    test_plaintext = "This is sensitive data that needs encryption!"
    print(f"\nTesting encryption/decryption:")
    print(f"Original: {test_plaintext}")
    
    encrypted = encrypt_data(test_plaintext)
    print(f"Encrypted: {json.dumps(encrypted, indent=2)}")
    
    decrypted = decrypt_data(encrypted)
    print(f"Decrypted: {decrypted}")
    print(f"Match: {test_plaintext == decrypted}")
    
    # Start Flask app
    print("\nStarting Flask server on http://localhost:5000")
    app.run(debug=True, host="127.0.0.1", port=5000)